Unexpected whitespace error

I used Sybil to test a simple docstring, but got an unexpected error.

**conftest.py**
```python
from sybil import Sybil
from sybil.parsers.rest import DocTestParser

pytest_collect_file = Sybil(
    parsers=[DocTestParser()],
    patterns=['*.py'],
).pytest()
```

**module.py**
```python
def f():
    """Do something.

    Example:
        >>> x = 3
        >>> x
        3
    """
    pass
```
**Traceback**
```
❯ pytest
================================================================================================ test session starts ================================================================================================
platform darwin -- Python 3.10.9, pytest-7.2.2, pluggy-1.0.0
rootdir: /Users/balaji/Documents/sybil-example, configfile: pytest.ini
collected 0 items / 1 error                                                                                                                                                                                         

====================================================================================================== ERRORS =======================================================================================================
____________________________________________________________________________________________ ERROR collecting module.py _____________________________________________________________________________________________
.venv/lib/python3.10/site-packages/sybil/integration/pytest.py:104: in collect
    self.document = self.sybil.parse(Path(self.fspath.strpath))
.venv/lib/python3.10/site-packages/sybil/sybil.py:143: in parse
    return type_.parse(str(path), *self.parsers, encoding=self.encoding)
.venv/lib/python3.10/site-packages/sybil/document.py:54: in parse
    for region in parser(document):
.venv/lib/python3.10/site-packages/sybil/parsers/abstract/doctest.py:41: in __call__
    source, options, want, exc_msg = self._parse_example(m, name, lineno)
/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/doctest.py:721: in _parse_example
    self._check_prefix(want_lines, ' '*indent, name,
/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/doctest.py:806: in _check_prefix
    raise ValueError('line %r of the docstring for %s has '
E   ValueError: line 8 of the docstring for /Users/balaji/Documents/sybil-example/module.py has inconsistent leading whitespace: '    """'
============================================================================================== short test summary info ==============================================================================================
ERROR module.py - ValueError: line 8 of the docstring for /Users/balaji/Documents/sybil-example/module.py has inconsistent leading whitespace: '    """'
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
================================================================================================= 1 error in 0.05s ==================================================================================================
```
**Dependencies**
```
❯ pip freeze
attrs==22.2.0
exceptiongroup==1.1.0
iniconfig==2.0.0
packaging==23.0
pluggy==1.0.0
pytest==7.2.2
sybil==4.0.1
tomli==2.0.1
```

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -11,6 +11,8 @@ common: &common
           parameters:
             image:
               - cimg/python:3.6
+              - cimg/python:3.7
+              - cimg/python:3.8
               - cimg/python:3.9
               - cimg/python:3.11
             extra_packages:

--- a/conftest.py
+++ b/conftest.py
@@ -1,4 +1,10 @@
 from doctest import ELLIPSIS
+from functools import lru_cache
+from pathlib import Path
+from typing import Tuple, List
+
+import pytest
+
 from sybil import Sybil
 from sybil.parsers.rest import (
     CaptureParser,
@@ -16,3 +22,23 @@ pytest_collect_file = Sybil(
     ],
     patterns=['*.rst', '*.py'],
 ).pytest()
+
+
+def _find_python_files() -> List[Tuple[Path, str]]:
+    paths = []
+    for path in Path(__file__).parent.rglob('*.py'):
+        source = Path(path).read_text()
+        paths.append((path, source))
+    return paths
+
+
+@pytest.fixture
+def all_python_files():
+    return _find_python_files()
+
+
+def pytest_generate_tests(metafunc):
+    files = _find_python_files()
+    ids = [str(f[0]) for f in files]
+    if "python_file" in metafunc.fixturenames:
+        metafunc.parametrize("python_file", files, ids=ids)

--- a/docs/api.rst
+++ b/docs/api.rst
@@ -25,6 +25,9 @@ Documents
 .. autoclass:: sybil.document.PythonDocument
   :members:
 
+.. autoclass:: sybil.document.PythonDocStringDocument
+  :members:
+
 Regions
 -------
 

--- a/sybil/document.py
+++ b/sybil/document.py
@@ -1,12 +1,15 @@
+import ast
 import re
+from ast import AsyncFunctionDef, FunctionDef, ClassDef, Module, Expr, Str, Constant
 from bisect import bisect
 from io import open
 from pathlib import Path
-from typing import List, Iterable, Pattern, Tuple, Match, Optional
+from typing import List, Iterable, Pattern, Tuple, Match, Optional, Sequence
 
 from .example import Example
 from .python import import_path
 from .region import Region
+from .text import LineNumberOffsets
 from .typing import Parser, Evaluator
 
 
@@ -132,6 +135,9 @@ class Document:
             yield start_match, end_match, source
 
 
+DOCSTRING_PUNCTUATION = re.compile('["\']{3}|["\']')
+
+
 class PythonDocument(Document):
     """
     A :class:`~sybil.Document` type that imports the document's source
@@ -149,3 +155,53 @@ class PythonDocument(Document):
         result = example.region.evaluator(example)
         self.evaluator = None
         return result
+
+
+class PythonDocStringDocument(PythonDocument):
+    """
+    A :class:`~sybil.document.PythonDocument` subclass that only considers the text of
+    docstrings in the document's source.
+    """
+
+    @staticmethod
+    def extract_docstrings(python_source_code: str) -> Sequence[Tuple[int, int, str]]:
+        line_offsets = LineNumberOffsets(python_source_code)
+        for node in ast.walk(ast.parse(python_source_code)):
+            if not isinstance(node, (AsyncFunctionDef, FunctionDef, ClassDef, Module)):
+                continue
+            if not (node.body and isinstance(node.body[0], Expr)):
+                continue
+            docstring = node.body[0].value
+            if not (
+                    isinstance(docstring, Str) or
+                    isinstance(docstring, Constant) and isinstance(docstring.value, str)
+            ):
+                continue
+            node_start = line_offsets.get(docstring.lineno-1, docstring.col_offset)
+            node_end = line_offsets.get(docstring.end_lineno-1, docstring.end_col_offset)
+            punc = DOCSTRING_PUNCTUATION.match(python_source_code, node_start, node_end)
+            punc_size = punc.end() - punc.start()
+            start = punc.end()
+            end = node_end - punc_size
+            yield start, end, python_source_code[start:end]
+
+    @classmethod
+    def parse(cls, path: str, *parsers: Parser, encoding: str = 'utf-8') -> 'Document':
+        """
+        Read the text from the supplied path to a Python source file and parse any docstrings
+        it contains into a document using the supplied parsers.
+        """
+        with open(path, encoding=encoding) as source:
+            document = cls(source.read(), path)
+            for start, end, text in cls.extract_docstrings(document.text):
+                docstring_document = cls(text, path)
+                print(repr(docstring_document.text))
+                print(docstring_document.text)
+                for parser in parsers:
+                    for region in parser(docstring_document):
+                        region.start += start
+                        region.end += start
+                        print(region)
+                        print(repr(document.text[region.start:region.end]))
+                        document.add(region)
+        return document

--- a/sybil/parsers/abstract/doctest.py
+++ b/sybil/parsers/abstract/doctest.py
@@ -14,7 +14,7 @@ class DocTestStringParser(BaseDocTestParser):
     the doctest example's source and the file name that the example came from.
     """
 
-    def __init__(self, evaluator: DocTestEvaluator):
+    def __init__(self, evaluator: DocTestEvaluator = DocTestEvaluator()):
         #: The evaluator to use for any doctests found in the supplied source string.
         self.evaluator: DocTestEvaluator = evaluator
 

--- a/sybil/parsers/abstract/doctest.py
+++ b/sybil/parsers/abstract/doctest.py
@@ -27,11 +27,6 @@ class DocTestStringParser(BaseDocTestParser):
         Each section starting with a ``>>>`` will form a separate region.
         """
         # a cut down version of doctest.DocTestParser.parse:
-        # If all lines begin with the same indentation, then strip it.
-        min_indent = self._min_indent(string)
-        if min_indent > 0:
-            string = '\n'.join([l[min_indent:] for l in string.split('\n')])
-
         charno, lineno = 0, 0
         # Find all doctest examples in the string:
         for m in self._EXAMPLE_RE.finditer(string):
@@ -47,7 +42,7 @@ class DocTestStringParser(BaseDocTestParser):
                     m.end(),
                     DocTestExample(source, want, exc_msg,
                                    lineno=lineno,
-                                   indent=min_indent + len(m.group('indent')),
+                                   indent=len(m.group('indent')),
                                    options=options),
                     self.evaluator
 

--- a/sybil/parsers/abstract/lexers.py
+++ b/sybil/parsers/abstract/lexers.py
@@ -5,6 +5,9 @@ from typing import Dict, Iterable, Pattern
 from sybil import Document
 from sybil.region import LexedRegion, Lexeme
 
+class LexingException(Exception):
+    pass
+
 
 class BlockLexer:
     """
@@ -52,6 +55,11 @@ class BlockLexer:
                 prefix=prefix, len_prefix=len(prefix)
             ))
             end_match = end_pattern.search(document.text, source_start)
+            if end_match is None:
+                raise LexingException(
+                    f'Could not match {end_pattern.pattern!r} in {document.path}:\n'
+                    f'{document.text[source_start:]!r}'
+                )
             source_end = end_match.start()
             source = document.text[source_start:source_end]
             lines = source.splitlines(keepends=True)

--- a/sybil/sybil.py
+++ b/sybil/sybil.py
@@ -2,13 +2,14 @@ import sys
 from pathlib import Path
 from typing import Sequence, Callable, Collection, Mapping, Optional, Type, List
 
-from .document import Document, PythonDocument
+from .document import Document, PythonDocStringDocument, PythonDocument
 from .typing import Parser
 
+PY37_AND_EARLIER = sys.version_info[:2] <= (3, 7)
 
 DEFAULT_DOCUMENT_TYPES = {
     None: Document,
-    '.py': PythonDocument,
+    '.py': PythonDocument if PY37_AND_EARLIER else PythonDocStringDocument,
 }
 
 

--- a/sybil/text.py
+++ b/sybil/text.py
@@ -1,0 +1,18 @@
+import re
+
+NEWLINE = re.compile("\n")
+
+
+class LineNumberOffsets:
+
+    def __init__(self, text: str):
+        self.offsets = {
+            line: match.start()+1 for (line, match) in enumerate(NEWLINE.finditer(text), start=1)
+        }
+        self.offsets[0] = 0
+
+    def get(self, line: int, column: int) -> int:
+        """
+        Return the character offset of the  zero based line number and column offset.
+        """
+        return self.offsets[line] + column

--- a/tests/functional/package_and_docs/src/parent/child/module_b.py
+++ b/tests/functional/package_and_docs/src/parent/child/module_b.py
@@ -5,6 +5,4 @@ def foo(text):
     """
     >>> parent_init('module_a.foo(parent_init)')
     'module_a.foo(parent_init:module_a.foo(parent_init))'
-
-    Need space after doctest examples!
     """

--- a/tests/helpers.py
+++ b/tests/helpers.py
@@ -1,3 +1,4 @@
+import ast
 import sys
 from contextlib import contextmanager
 from os.path import dirname, join
@@ -6,9 +7,10 @@ from shutil import copytree
 from tempfile import NamedTemporaryFile
 from textwrap import dedent
 from traceback import TracebackException
-from typing import Tuple, List
+from typing import Tuple, List, Sequence
 from unittest import TextTestRunner, main as unittest_main
 
+import pytest
 from _pytest._code import ExceptionInfo
 from _pytest.capture import CaptureFixture
 from _pytest.config import main as pytest_main
@@ -233,3 +235,19 @@ def add_to_python_path(path: Path):
         sys.path.append(str(path))
         yield
         sys.path.pop()
+
+
+def ast_docstrings(python_source_code: str) -> Sequence[str]:
+    for node in ast.walk(ast.parse(python_source_code)):
+        try:
+            docstring = ast.get_docstring(node, clean=False)
+        except TypeError:
+            pass
+        else:
+            if docstring:
+                yield docstring
+
+
+def skip_if_37_or_older():
+    return pytest.mark.skipif(sys.version_info[:2] < (3, 8), reason="requires python3.8 or higher")
+

--- a/tests/samples/comments.py
+++ b/tests/samples/comments.py
@@ -1,0 +1,4 @@
+# XXXX 4 check
+"""
+YYY 3 check
+"""

--- a/tests/samples/lexing-fail.txt
+++ b/tests/samples/lexing-fail.txt
@@ -1,0 +1,2 @@
+START
+EDN

--- a/tests/test_doctest.py
+++ b/tests/test_doctest.py
@@ -1,13 +1,15 @@
 # coding=utf-8
-from doctest import REPORT_NDIFF, ELLIPSIS
+from doctest import REPORT_NDIFF, ELLIPSIS, DocTestParser as BaseDocTestParser
+from pathlib import Path
 
 import pytest
 from testfixtures import compare
 
-from sybil.document import Document
+from sybil.document import Document, PythonDocStringDocument
 from sybil.example import SybilFailure
+from sybil.parsers.abstract import DocTestStringParser
 from sybil.parsers.rest import DocTestParser, DocTestDirectiveParser
-from tests.helpers import sample_path, parse
+from .helpers import sample_path, parse, FUNCTIONAL_TEST_DIR, skip_if_37_or_older
 
 
 def test_pass():
@@ -119,3 +121,64 @@ def test_directive_with_options():
         "    - Unexpected!\n"
         "    + 2\n"
     ))
+
+
+# Number of doctests that can't be parsed in a file when looking at the whole file source:
+ROOT = Path(FUNCTIONAL_TEST_DIR)
+UNPARSEABLE = {
+   ROOT / 'package_and_docs' / 'src' / 'parent' / 'child' / 'module_b.py': 1,
+}
+MINIMUM_EXPECTED_DOCTESTS = 9
+
+
+@skip_if_37_or_older()
+def test_sybil_example_count(all_python_files):
+    parser = DocTestStringParser()
+
+    seen_examples_from_source = 0
+    seen_examples_from_docstrings = 0
+
+    for path, source in all_python_files:
+        seen_examples_from_source += len(tuple(parser(source, path)))
+        for start, end, docstring in PythonDocStringDocument.extract_docstrings(source):
+            seen_examples_from_docstrings += len(tuple(parser(docstring, path)))
+
+    assert seen_examples_from_source > MINIMUM_EXPECTED_DOCTESTS, seen_examples_from_source
+    assert seen_examples_from_docstrings > MINIMUM_EXPECTED_DOCTESTS, seen_examples_from_docstrings
+    assert seen_examples_from_docstrings == seen_examples_from_source
+
+
+def check_sybil_against_doctest(path, text):
+    skip_if_37_or_older()
+    problems = []
+    name = str(path)
+    regions = list(DocTestStringParser()(text, path))
+    sybil_examples = [region.parsed for region in regions]
+    doctest_examples = BaseDocTestParser().get_examples(text, name)
+    problems.append(compare(
+        expected=doctest_examples, actual=sybil_examples, raises=False, show_whitespace=True
+    ))
+    for region in regions:
+        example = region.parsed
+        region_source = text[region.start:region.end]
+        for name in 'source', 'want':
+            expected = getattr(example, name)
+            if expected not in region_source:  # pragma: no cover - Only on failures!
+                problems.append(f'{region}:{name}\n{expected!r} not in {region_source!r}')
+    problems = [problem for problem in problems if problem]
+    if problems:  # pragma: no cover - Only on failures!
+        raise AssertionError('doctests not correctly extracted:\n\n'+'\n\n'.join(problems))
+
+
+def test_all_docstest_examples_extracted_from_source_correctly(python_file):
+    path, source = python_file
+    if path in UNPARSEABLE:
+        return
+    check_sybil_against_doctest(path, source)
+
+
+@skip_if_37_or_older()
+def test_all_docstest_examples_extracted_from_docstrings_correctly(python_file):
+    path, source = python_file
+    for start, end, docstring in PythonDocStringDocument.extract_docstrings(source):
+        check_sybil_against_doctest(path, docstring)

--- a/tests/test_document.py
+++ b/tests/test_document.py
@@ -1,0 +1,40 @@
+"""
+This is a module doc string.
+"""  # This is a comment on how annoying 3.7 and earlier are!
+from functools import partial
+from pathlib import Path
+
+from testfixtures import compare
+
+from sybil.document import PythonDocStringDocument
+from .helpers import ast_docstrings, skip_if_37_or_older
+
+
+@skip_if_37_or_older()
+def test_extract_docstring():
+    """
+    This is a function docstring.
+    """
+    python_source_code = Path(__file__).read_text()
+    expected = list(ast_docstrings(python_source_code))
+    actual = list(PythonDocStringDocument.extract_docstrings(python_source_code))
+    compare(expected, actual=[text for _, _, text in actual], show_whitespace=True)
+    compare(expected, actual=[python_source_code[s:e] for s, e, _ in actual], show_whitespace=True)
+
+
+@skip_if_37_or_older()
+def test_all_docstrings_extracted_correctly(python_file):
+    problems = []
+    path, source = python_file
+    expected = list(ast_docstrings(source))
+    actual = list(PythonDocStringDocument.extract_docstrings(source))
+    check = partial(compare, prefix=str(path), raises=False, show_whitespace=True)
+    for check_result in (
+        check(expected=len(expected), actual=len(actual)),
+        check(expected, actual=[text for _, _, text in actual]),
+        check(expected, actual=[source[s:e] for s, e, _ in actual]),
+    ):
+        if check_result:  # pragma: no cover - Only on failures!
+            problems.append(check_result)
+    if problems:  # pragma: no cover - Only on failures!
+        raise AssertionError('docstrings not correctly extracted:\n\n'+'\n\n'.join(problems))

--- a/tests/test_functional.py
+++ b/tests/test_functional.py
@@ -7,7 +7,7 @@ from pytest import CaptureFixture
 from sybil.python import import_cleanup
 from .helpers import (
     run_pytest, run_unittest, PYTEST, run, write_config, UNITTEST, write_doctest,
-    functional_sample, clone_functional_sample
+    functional_sample, clone_functional_sample, skip_if_37_or_older
 )
 
 
@@ -331,6 +331,7 @@ def test_modules_not_importable_unittest(tmpdir: local, capsys: CaptureFixture[s
     out.then_find("ModuleNotFoundError: No module named 'b'")
 
 
+@skip_if_37_or_older()
 @pytest.mark.parametrize('runner', [PYTEST, UNITTEST])
 def test_package_and_docs(tmpdir: local, capsys: CaptureFixture[str], runner: str):
     root = clone_functional_sample('package_and_docs', tmpdir)

--- a/tests/test_helpers.py
+++ b/tests/test_helpers.py
@@ -5,7 +5,7 @@ import pytest
 from py.path import local
 
 from sybil.python import import_cleanup
-from .helpers import Finder
+from .helpers import Finder, ast_docstrings
 
 
 def test_finder_present_but_is_not_present():
@@ -36,3 +36,15 @@ def test_import_cleanup(tmpdir: local):
 
     assert sys.modules == initial_modules
     assert sys.path == initial_path
+
+
+def test_all_python_files(all_python_files):
+    count = len(all_python_files)
+    assert count > 50, count
+
+
+def test_ast_docstrings(all_python_files):
+    seen_docstrings = 0
+    for _, source in all_python_files:
+        seen_docstrings += len(tuple(ast_docstrings(source)))
+    assert seen_docstrings > 50, seen_docstrings

--- a/tests/test_lexing.py
+++ b/tests/test_lexing.py
@@ -1,0 +1,13 @@
+import re
+
+from testfixtures import ShouldRaise
+
+from sybil.parsers.abstract.lexers import BlockLexer, LexingException
+from .helpers import lex, sample_path
+
+
+def test_examples_from_parsing_tests():
+    lexer = BlockLexer(start_pattern=re.compile('START'), end_pattern_template='END')
+    path = sample_path('lexing-fail.txt')
+    with ShouldRaise(LexingException(f"Could not match 'END' in {path}:\n'\\nEDN\\n'")):
+        lex('lexing-fail.txt', lexer)

--- a/tests/test_sybil.py
+++ b/tests/test_sybil.py
@@ -9,7 +9,7 @@ import pytest
 from py.path import local
 
 from sybil import Sybil, Region
-from sybil.document import Document
+from sybil.document import Document, PythonDocument
 from sybil.example import Example, SybilFailure
 
 from .helpers import sample_path, write_doctest
@@ -251,6 +251,17 @@ class TestSybil:
         assert type(document) is TextDocument
         document = sybil.parse(write_doctest(tmpdir, 'test.rst'))
         assert type(document) is Document
+
+    def test_override_document_mapping(self):
+        sybil = Sybil([parse_for_x, parse_for_y], document_types={'.py': PythonDocument})
+        document = sybil.parse(Path(sample_path('sample1.txt')))
+        assert (evaluate_examples(document) ==
+                ['X count was 4, as expected',
+                 'Y count was 3, as expected'])
+        document = sybil.parse(Path(sample_path('comments.py')))
+        assert (evaluate_examples(document) ==
+                ['X count was 4, as expected',
+                 'Y count was 3, as expected'])
 
     def test_addition(self):
         rest = Sybil([parse_for_x])

